### PR TITLE
Add type hinting

### LINKED ISSUE #32: Add hints to functions
Add type hints to all function parameters and outputs

### PR DESCRIPTION
**Pull request by FrancescoVaiani**
Add type hinting to functions and classes

Close #32 

Signed-off-by: Francesco Vaiani <francesco.vaiani@secomind.com>

### CODE REVIEW

File: astarte/device/device.py [hunk 1]
```diff
@@ -92,22 +95,22 @@ def __init__(self,
...
-
-        self.on_connected = None
-        self.on_disconnected = None
+        self.__device_id: str = device_id
+        self.__realm: str = realm
+        self.__pairing_base_url: str = pairing_base_url
+        self.__persistency_dir: str = persistency_dir
+        self.__credentials_secret: str = credentials_secret
+        self.__jwt_token: Optional[str] = None
+        self.__is_crypto_setup: bool = False
```
Comment (sorru94) [inline astarte/device/device.py:104]: I see that in some places the type hinting is used, in others it's not.
Is this PR proposing to use it only where needed, or would it place it everywhere?
On some of these, like the `bool` the type is already clear. Using type hinting there might just make uncecessary clutter.

File: astarte/device/introspection.py [hunk 2]
```diff
@@ -25,9 +26,9 @@ class Introspection:
     name but different major/minor are not supported.
     """
     def __init__(self):
-        self.__interfaces_list = {}
+        self.__interfaces_list: dict = {}
```
Comment (sorru94) [inline astarte/device/introspection.py:29]:
```suggestion
        self.__interfaces_list = {}
```
I think it's kind of implied by the initialization.

File: astarte/device/introspection.py [hunk 3]
```diff
@@ -83,9 +84,9 @@ def get_all_interfaces(self):
         list
             The list of all Interfaces in the Introspection
         """
-        return self.__interfaces_list.values()
+        return list(self.__interfaces_list.values())
```
Comment (sorru94) [inline astarte/device/introspection.py:87]:
```suggestion
        return [self.__interfaces_list.values()]
```

Comment (FrancescoVaiani) [reply astarte/device/introspection.py]: `list()` here is not used to **make** a list, but to make clear that the output value is of the `list` type. 
The function `dict.values()` outputs a `_dict_values` that is a disguised list, so no square brackets are needed.

**Review by sorru94 — APPROVED** [review]

File: astarte/device/device.py [hunk 4]
```diff
@@ -283,10 +287,10 @@ def send(self, interface_name, interface_path, payload, timestamp=None):
...
 
     def send_aggregate(self,
-                       interface_name,
-                       interface_path,
-                       payload,
-                       timestamp=None):
+                       interface_name: str,
+                       interface_path: str,
+                       payload: dict,
+                       timestamp=None) -> None:
```
Comment (rbino) [inline astarte/device/device.py:293]: Missing hinting on `timestamp`

File: astarte/device/device.py [hunk 5]
```diff
@@ -97,17 +100,17 @@ def __init__(self,
...
         self.__introspection = Introspection()
         self.__is_connected = False
         self.__loop = loop
         self.__ignore_ssl_errors = ignore_ssl_errors
 
-        self.on_connected = None
-        self.on_disconnected = None
+        self.on_connected: Optional[Callable[[Device], None]] = None
+        self.on_disconnected: Optional[Callable[[Device, int], None]] = None
         self.on_aggregate_data_received = None
```
Comment (rbino) [inline astarte/device/device.py:112]: Why this callback does not have hinting?

**Review by rbino — CHANGES_REQUESTED** [review]

File: astarte/device/device.py (hunk 5 above)
Comment (FrancescoVaiani) [reply astarte/device/device.py:112]: This callback is not actually used inside all the SDK and is planned to be removed in a separate PR

**Review by rbino — APPROVED** [review]

**Pull request merged**